src/main/java/me/magnet/consultant/RoutingStrategies.java

```diff
@@ -0,0 +1,110 @@
+package me.magnet.consultant;
+
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+public class RoutingStrategies {
+
+	public static final RoutingStrategy RANDOMIZED_WEIGHTED_DISTANCE = new RoutingStrategy() {
+
+		private final Random random = new Random();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> instances = serviceLocator.listInstances(serviceName);
+			if (instances.size() <= 1) {
+				// No routing to do.
+				return instances;
+			}
+
+			// Pick a random number between 0-1.
+			double chance = random.nextDouble();
+
+			/*
+			 * Pick the index of the first instance to try based on this chance.
+			 * If the chance was < 0.5 then use the instance on index 0.
+			 * If the chance was >= 0.5 and < 0.75 then use the instance on index 1.
+			 * If the chance was >= 0.75 and < 0.875 then use the instance on index 2.
+			 * If the chance was >= 0.875 and < 0.9375 then use the instance on index 3.
+			 * Etc...
+			 */
+			int index = 0;
+			double threshold = 0.5;
+			while (threshold < chance && index < instances.size() - 1) {
+				index++;
+				threshold = Math.pow(0.5, index + 1);
+			}
+
+			return splice(instances, index);
+		}
+	};
+
+	public static final RoutingStrategy ROUND_ROBIN = new RoutingStrategy() {
+
+		private final Multimap<String, ServiceInstance> lastRequested = HashMultimap.create();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> nextAvailable = serviceLocator.listInstances(serviceName);
+			Set<ServiceInstance> lastRequests = Sets.newHashSet(lastRequested.get(serviceName));
+
+			// If all instances have been hit at least once, reset the counter for this service.
+			if (lastRequests.containsAll(nextAvailable)) {
+				lastRequested.removeAll(serviceName);
+			}
+
+			// List the instances to hit in order of not hit to already hit (conserving network distance).
+			List<ServiceInstance> nextUp = nextAvailable.stream()
+					.filter(next -> !lastRequests.contains(next))
+					.collect(Collectors.toList());
+			nextAvailable.stream()
+					.filter(lastRequests::contains)
+					.forEachOrdered(nextUp::add);
+
+			// Register that nextUp[0] instance is going to get hit first.
+			if (!nextUp.isEmpty()) {
+				lastRequested.put(serviceName, nextUp.get(0));
+			}
+
+			return nextUp;
+		}
+	};
```

Comment:
> Returning one item is already done by the locate() method in Consultant

Ahh I see, I missed that. Maybe it would then make sense to deprecate the `list` functions, since I think the problems I outlined above are fundamental, and can't be solved (even with iterators, like you said). The issue with `locate()` and followers/racing can also be solved in different ways that don't have the downside of lists (e.g. use a blacklist, or introduce a concept of sessions/transactions that do some bookkeeping). 

> The current implementation only guarantees that every service will be hit equally often, it doesn't make any guarantees about ordering.

- It doesn't, because of the concerns above
- If that's the only guarantee it makes then it is not round robin, and then it is for mosts intents and purposes no better than random (which also has that guarantee). With the low number of instances for most of our services it will actually behaves very, very similar to random if this is not fixed. 

I really think you should fix this part of round robin, because without this it's not round robin and frankly not a particularly useful strategy (compared to randomize).